qiskit/pulse/parameter_manager.py

```diff
@@ -300,18 +300,19 @@ def visit_ScheduleBlock(self, node: ScheduleBlock):
 
         .. note:: ``ScheduleBlock`` can have parameters in blocks and its alignment.
         """
-        for parameter in node.parameters:
-            self.parameters.add(parameter)
+        # Note that node.parameters returns parameters of main program with subroutines.
+        # The manager of main program is not aware of parameters in subroutines.
+        self.parameters = self.parameters | node._parameter_manager.parameters
 
     def visit_Schedule(self, node: Schedule):
         """Visit ``Schedule``. Recursively visit schedule children and search parameters."""
-        for parameter in node.parameters:
-            self.parameters.add(parameter)
+        self.parameters = self.parameters | node.parameters
 
     def visit_AlignmentKind(self, node: AlignmentKind):
         """Get parameters from block's ``AlignmentKind`` specification."""
         for param in node._context_params:
-            self.visit(param)
+            if isinstance(param, ParameterExpression):
+                self.parameters = self.parameters | param.parameters
```

Comment:
Is there any chance you might need to recurse into `visit(param)` if it's no a `ParameterExpression`?

Reply:
Not for now. User can still add custom alignment by subclassing `AlignmentKind` and such a custom alignment could take some object which is not `ParameterExpression`. But I think I've never come up with such fancy alignment logic.